File: Inboxes/Node.js/Web/Plain/src/storage/adapters/SQLite.js
const DatabaseAdapter = require('../DatabaseAdapter');

class SQLiteAdapter extends DatabaseAdapter {
  constructor(defaultDB) {
    super();
    this.defaultDB = defaultDB || 'data';
  }

  async connect() {
    const Database = require('better-sqlite3');
    this.db = new Database(process.env.DB_FILE || `${this.defaultDB}.db`);
    this.db.pragma('journal_mode = WAL');
  }

  _colType(col) {
    switch (col.type) {
      case 'string': case 'text': return 'TEXT';
      case 'integer': return 'INTEGER';
      case 'boolean': return 'INTEGER';
      case 'datetime': return 'TEXT';
      default: return 'TEXT';
    }
  }

  _colDef(col) {
    let def = `${col.name} ${this._colType(col)}`;
    if (col.required) def += ' NOT NULL';
    if (col.default !== undefined) {
      if (col.default === 'now') def += " DEFAULT (datetime('now'))";
      else if (typeof col.default === 'string') def += ` DEFAULT '${col.default}'`;
      else def += ` DEFAULT ${col.default}`;
    }
    return def;
  }

  async init(table, columns) {
    const cols = columns.map(c => this._colDef(c)).join(', ');
    this.db.exec(`CREATE TABLE IF NOT EXISTS ${table} (id INTEGER PRIMARY KEY AUTOINCREMENT, ${cols})`);
  }

  async getAll(table, options = {}) {
    let sql = `SELECT * FROM ${table}`;
    if (options.orderBy) sql += ` ORDER BY ${options.orderBy} ${options.orderDir || 'ASC'}`;
    if (options.limit) sql += ` LIMIT ${options.limit}`;
    return this.db.prepare(sql).all();
  }

  async getById(table, id) {
    return this.db.prepare(`SELECT * FROM ${table} WHERE id = ?`).get(id) || null;
  }

  async create(table, data) {
    const keys = Object.keys(data);
    const vals = Object.values(data);
    const placeholders = keys.map(() => '?').join(', ');
    const info = this.db.prepare(`INSERT INTO ${table} (${keys.join(', ')}) VALUES (${placeholders})`).run(...vals);
    return this.db.prepare(`SELECT * FROM ${table} WHERE id = ?`).get(info.lastInsertRowid);
  }

  async update(table, id, data) {
    const keys = Object.keys(data);
    const vals = Object.values(data);
    const setClause = keys.map(k => `${k} = ?`).join(', ');
    vals.push(id);
    this.db.prepare(`UPDATE ${table} SET ${setClause} WHERE id = ?`).run(...vals);
    return this.db.prepare(`SELECT * FROM ${table} WHERE id = ?`).get(id);
  }

  async delete(table, id) {
    const info = this.db.prepare(`DELETE FROM ${table} WHERE id = ?`).run(id);
    return { changes: info.changes };
  }

  async clear(table) {
    const info = this.db.prepare(`DELETE FROM ${table}`).run();
    return { changes: info.changes };
  }

  async search(table, field, query) {
    return this.db.prepare(`SELECT * FROM ${table} WHERE LOWER(${field}) LIKE ? ORDER BY id ASC`).all(`%${query.toLowerCase()}%`);
  }

  async close() {
    this.db.close();
  }
}
module.exports = SQLiteAdapter;

File: TasksList/Python/Web/Plain/src/storage/adapters/sqlite.py
import os
import sqlite3
from storage.database_adapter import DatabaseAdapter

TYPE_MAP = {
    "TEXT": "TEXT",
    "INTEGER": "INTEGER",
    "TIMESTAMP": "TEXT",
}

class SQLite(DatabaseAdapter):
    def __init__(self):
        self.conn = None
        self.connect()

    def connect(self):
        if self.conn:
            return
        self.conn = sqlite3.connect(os.getenv("DB_FILE", "data.db"))
        self.conn.row_factory = sqlite3.Row

    def _translate(self, col_def):
        for generic, specific in TYPE_MAP.items():
            col_def = col_def.replace(generic, specific)
        col_def = col_def.replace("DEFAULT NOW()", "DEFAULT (datetime('now'))")
        return col_def

    def init_table(self, table, columns=None):
        self.connect()
        cur = self.conn.cursor()
        cols = ["id INTEGER PRIMARY KEY AUTOINCREMENT"]
        if columns:
            for name, col_def in columns.items():
                cols.append(f"{name} {self._translate(col_def)}")
        cur.execute(f"CREATE TABLE IF NOT EXISTS {table} ({', '.join(cols)})")
        self.conn.commit()
        cur.close()

    def _query(self, sql, params=None, fetchone=False, fetchall=False):
        self.connect()
        cur = self.conn.cursor()
        cur.execute(sql, params or [])
        if fetchone:
            row = cur.fetchone()
            cur.close()
            return dict(row) if row else None
        if fetchall:
            rows = [dict(r) for r in cur.fetchall()]
            cur.close()
            return rows
        cur.close()

    def _execute(self, sql, params=None):
        self.connect()
        cur = self.conn.cursor()
        cur.execute(sql, params or [])
        self.conn.commit()
        rowcount = cur.rowcount
        cur.close()
        return rowcount

    def get_all(self, table):
        return self._query(f"SELECT * FROM {table} ORDER BY id ASC", fetchall=True)

    def get_by_id(self, table, id):
        return self._query(f"SELECT * FROM {table} WHERE id = ?", (id,), fetchone=True)

    def create(self, table, data):
        cols = ", ".join(data.keys())
        placeholders = ", ".join("?" for _ in data)
        values = list(data.values())
        self.connect()
        cur = self.conn.cursor()
        cur.execute(f"INSERT INTO {table} ({cols}) VALUES ({placeholders})", values)
        self.conn.commit()
        cur.execute(f"SELECT * FROM {table} WHERE id = last_insert_rowid()")
        row = dict(cur.fetchone())
        cur.close()
        return row

    def update(self, table, id, data):
        set_clause = ", ".join(f"{k} = ?" for k in data)
        values = list(data.values()) + [id]
        return self._execute(f"UPDATE {table} SET {set_clause} WHERE id = ?", values) > 0

    def delete(self, table, id):
        return self._execute(f"DELETE FROM {table} WHERE id = ?", (id,)) > 0

    def search(self, table, field, query):
        return self._query(
            f"SELECT * FROM {table} WHERE LOWER({field}) LIKE ? ORDER BY id ASC",
            (f"%{query.lower()}%",), fetchall=True
        )

    def close(self):
        if self.conn:
            self.conn.close()
